---
title: "Error example"
output:
  #rmarkdown::html_vignette:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{Error example}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```R
library(gwhap)
```

# Difference between both 1000 genome interpolated map and snp physical position in term of rsid, position

get the snp physical position
```R
snp_physical_positions = "/neurospin/ukb/genetic/GENETIC_DATA_500k/HAPLOTYPES/v2/ukb_hap_"
snp_list = c()
for (chr in 1:23){
  bgi_file_path = sprintf('%schr%s_v2.bgen.bgi', snp_physical_positions, chr)
  # check if file path exist
  if(file.exists(bgi_file_path)){
    # get bgi file
    tmp = get_bgi_file(bgi_file_path)
    # add chromosome code
    tmp$chromosome = chr
    # change column name for correspondance with bim file
    colnames(tmp) <- c('chromosome', 'bp', 'snp', 'number_of_alleles', 'allele1', 'allele2', 'file_start_position', 'size_in_bytes')
    # concatenation
    snp_list  = rbind(snp_list, tmp)
  }
}
```

get the 1000 genome interpolated map

```R
m1000_genome_map = c()
m1000_genome_map_path = "/neurospin/brainomics/bio_resources/hg19_maps/interpolated_maps/1000-genomes-genetic-maps-master/interpolated_from_hapmap"
for (chr in unique(snp_list$chromosome)){
  m1000_genome_map = rbind(m1000_genome_map, get_1000_genome_interpolated_map(sprintf('%s/chr%s.interpolated_genetic_map.gz', m1000_genome_map_path, chr)))
}
```

```R
nrow(m1000_genome_map)
```
==> 38.219.238

```R
nrow(snp_list)
```
==> 658.720

```R
common_snp = m1000_genome_map[m1000_genome_map$rsid %in% snp_list$snp, ]
nrow(common_snp)
```
==> 642.385




# Create the augmented genetic map
```R
sprintf('Start time : %s', Sys.time())
snp_physical_positions = "/neurospin/ukb/genetic/GENETIC_DATA_500k/HAPLOTYPES/v2/ukb_hap_"
m1000_genome_map = "/neurospin/brainomics/bio_resources/hg19_maps/interpolated_maps/1000-genomes-genetic-maps-master/interpolated_from_hapmap"
output = '/neurospin/tmp/ymekki/new_repo/gwhap/augmented_genetic_map_1000_genome_interpolated'
create_augmented_genetic_map(snp_physical_positions=snp_physical_positions, genetic_map_dir=m1000_genome_map, map_name='1000_genome_interpolated', save_genetic_map=TRUE, output=output)
sprintf('End time : %s', Sys.time())
```

# debugging

## read the augmented genetic map created above
```R
augmented_genetic_map_dir = '/neurospin/tmp/ymekki/new_repo/gwhap/augmented_genetic_map_1000_genome_interpolated'
augmented_map_df = get_augmented_genetic_map(augmented_genetic_map_dir, chromosomes=1:23)
```

## Read the genetic map -S.K-
```R
load("/neurospin/brainomics/2019_gwhap/data/genMap_all.Rdata")
```


## merging both genetic map
```R
df_merged_by_rsid = merge(augmented_map_df, genMap_all, by.x = 'rsid', by.y = 'rsid')
```

```R
head(df_merged_by_rsid, 10)
```

|   |         rsid|     cM.x|    pos.x| chr.x|     cM.y|    pos.y| chr.y|
|:-:|:--------:   |:-------:|:-------:|:----:|:-------:|:-------:|:----:|
|1  |Affx-10000328| 67.88376| 39542062|     1| 67.88376| 39542062|  chr1|
|2  |Affx-10018171| 88.61191| 95705380|    13| 88.61191| 95705380| chr13|
|3  |Affx-10020391| 89.13778| 95859035|    13| 89.13778| 95859035| chr13|
|4  |Affx-10030637| 89.85192| 96638636|    13| 89.85192| 96638636| chr13|
|5  |Affx-10047176| 68.08397| 39917888|     1| 68.08397| 39917888|  chr1|
|6  |Affx-10053866| 91.47219| 98640318|    13| 91.47219| 98640318| chr13|
|7  |Affx-10055540| 91.75806| 98749827|    13| 91.75806| 98749827| chr13|
|8  |Affx-10058161| 92.30553| 98928242|    13| 92.30553| 98928242| chr13|
|9  |Affx-10065209| 93.59793| 99392176|    13| 93.59793| 99392176| chr13|
|10 |Affx-10067946| 93.88717| 99592598|    13| 93.88717| 99592598| chr13|



```R
tail(df_merged_by_rsid, 10)
```

|      |      rsid|       cM.x|     pos.x| chr.x|      cM.y|     pos.y| chr.y|
|:----:|:-----:   |:---------:|:--------:|:----:|:--------:|:--------:|:----:|
|658711| rs9999718|  99.46464 |  84762853|     4|  99.46464|  84762853|  chr4|
|658712| rs9999742|  83.91372 |  67943855|     4|  83.91372|  67943855|  chr4|
|658713| rs9999789| 119.85002 | 110510946|     4| 119.85002| 110510946|  chr4|
|658714| rs9999806|  56.18185 |  34952348|     4|  56.18185|  34952348|  chr4|
|658715| rs9999843| 190.57694 | 181531165|     4| 190.57694| 181531165|  chr4|
|658716| rs9999852| 172.85596 | 168403671|     4| 172.85596| 168403671|  chr4|
|658717|  rs999988|  61.67135 |  37635299|     9|  61.67135|  37635299|  chr9|
|658718|  rs999989|  28.00999 |   9658077|    18|  28.00999|   9658077| chr18|
|658719| rs9999927| 140.13134 | 135945570|     4| 140.13134| 135945570|  chr4|
|658720| rs9999971|  10.71974 |   5284714|     4|  10.71974|   5284714|  chr4|


```R
same <- df_merged_by_rsid[df_merged_by_rsid$cM.x == df_merged_by_rsid$cM.y, ]
nrow(same)
```
==> 653924


```R
diff <- df_merged_by_rsid[df_merged_by_rsid$cM.x != df_merged_by_rsid$cM.y, ]
nrow(diff)
```
==> 4798

```R
diff_values = diff$cM.x - diff$cM.y
mean(diff_values, na.rm=TRUE)
```
==> -1.844319e-16
```R
max(diff_values, na.rm=TRUE)
```

==> 5.115908e-13

```R
min(diff_values, na.rm=TRUE)
```
==> -5.115908e-13



```R
summary(df_merged_by_rsid)
```

|   rsid            |    cM.x         |   pos.x          |      chr.x     |       cM.y      |      pos.y         |       chr.y     |
|:-----------------:|:------------:   |:----------------:|:--------------:|:---------------:|:------------------:|:---------------:|
|   Length:658720   |   Min.   :  0.00| Min.   :     6157|   Min.   : 1.00| Min.   :  0.00  | Min.   :     6157  | chr1   : 53260  |
|   Class :character|   1st Qu.: 47.99| 1st Qu.: 30294386|   1st Qu.: 4.00| 1st Qu.: 47.98  | 1st Qu.: 30294386  |chr2   : 52631   |
|   Mode  :character|   Median : 83.15| Median : 66612348|   Median : 8.00| Median : 83.15  |Median : 66612348   |chr6   : 45886   |
|                   |   Mean   : 92.11| Mean   : 77478883|   Mean   : 9.01| Mean   : 92.11  |Mean   : 77478883   |chr3   : 44263   |
|                   |   3rd Qu.:128.54| 3rd Qu.:114812360|   3rd Qu.:13.00| 3rd Qu.:128.54  |3rd Qu.:114812360   |chr4   : 41213   |
|                   |   Max.   :286.28| Max.   :249222527|   Max.   :22.00| Max.   :286.28  |Max.   :249222527   |chr5   : 38734   |
|                   |   NA's   :2     |                  |                |                 |                    |(Other):382733   |




