<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Genome wide example</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Genome wide example</h1>


<div id="TOC">
<ul>
<li><a href="#download-the-rutgers-genetic-map">Download the Rutgers genetic map</a></li>
<li><a href="#create-the-augmented-genetic-map">Create the augmented genetic map</a></li>
<li><a href="#blocs-creation">Blocs creation</a></li>
<li><a href="#determine-the-haplotypes">Determine the haplotypes</a></li>
<li><a href="#test-the-haplotypes">Test the haplotypes</a></li>
</ul>
</div>

<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(gwhap)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"></a>
<a class="sourceLine" id="cb1-3" data-line-number="3">RESOURCE_ROOT =<span class="st"> &quot;&lt;YOUR_LOCAL_RESOURCE_FS&gt;&quot;</span></a></code></pre></div>
<div id="download-the-rutgers-genetic-map" class="section level1">
<h1>Download the Rutgers genetic map</h1>
<p>Download the Rutgers genetic map v3 (Kosambi)</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">download_rutgers_map</span>(RESOURCE_ROOT)</a></code></pre></div>
</div>
<div id="create-the-augmented-genetic-map" class="section level1">
<h1>Create the augmented genetic map</h1>
<p>Create the augmented genetic map using the rutgers genetic map v3</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">snp_physical_positions =<span class="st"> &quot;/neurospin/ukb/genetic/GENETIC_DATA_500k/HAPLOTYPES/v2/ukb_hap_&quot;</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">rutgers_map =<span class="st"> &quot;/neurospin/brainomics/bio_resources/rutgers_map_v3&quot;</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3">output =<span class="st"> '/neurospin/tmp/ymekki/new_repo/gwhap/augmented_genetic_map'</span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="kw">create_augmented_genetic_map</span>(<span class="dt">snp_physical_positions=</span>snp_physical_positions, <span class="dt">genetic_map_dir=</span>rutgers_map, <span class="dt">save_genetic_map=</span><span class="ot">TRUE</span>, <span class="dt">output=</span>output)</a></code></pre></div>
</div>
<div id="blocs-creation" class="section level1">
<h1>Blocs creation</h1>
<p>Create blocs with a single delta</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">augmented_genetic_map_dir =<span class="st"> '/neurospin/tmp/ymekki/new_repo/gwhap/augmented_genetic_map'</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">output =<span class="st"> '/neurospin/tmp/ymekki/new_repo/gwhap/blocs'</span></a>
<a class="sourceLine" id="cb4-3" data-line-number="3">augmented_map_df =<span class="st"> </span><span class="kw">get_augmented_genetic_map</span>(augmented_genetic_map_dir, <span class="dt">chromosomes=</span><span class="dv">1</span><span class="op">:</span><span class="dv">23</span>)</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">df_blocks =<span class="st"> </span><span class="kw">create_blocks</span>(augmented_map_df, <span class="dt">delta=</span><span class="fl">1e-3</span>, <span class="dt">save_blocs=</span><span class="ot">TRUE</span>, <span class="dt">output=</span>output)</a></code></pre></div>
</div>
<div id="determine-the-haplotypes" class="section level1">
<h1>Determine the haplotypes</h1>
<p>Set parameters :</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co"># get blocs</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">blocs_dir =<span class="st"> '/neurospin/tmp/ymekki/new_repo/gwhap/blocs'</span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3">df_blocs  =<span class="st"> </span><span class="kw">get_blocs</span>(blocs_dir)</a>
<a class="sourceLine" id="cb5-4" data-line-number="4"></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="co"># bgen file</span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6">bgen_file_path =<span class="st"> '/neurospin/ukb/genetic/GENETIC_DATA_500k/HAPLOTYPES/v2/ukb_hap'</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"></a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="co"># phenotype</span></a>
<a class="sourceLine" id="cb5-9" data-line-number="9">phenotype_path =<span class="st"> '/neurospin/ukb/workspace/pheno_boxcox/all_pheno_res.tsv'</span></a>
<a class="sourceLine" id="cb5-10" data-line-number="10">phenotype =<span class="st"> </span><span class="kw">read_delim</span>(phenotype_path, <span class="dt">delim=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</a>
<a class="sourceLine" id="cb5-11" data-line-number="11"></a>
<a class="sourceLine" id="cb5-12" data-line-number="12"><span class="co"># index path</span></a>
<a class="sourceLine" id="cb5-13" data-line-number="13">sample_file_path =<span class="st"> '/neurospin/ukb/genetic/GENETIC_DATA_500k/HAPLOTYPES/v2/ukb25251_hap'</span></a>
<a class="sourceLine" id="cb5-14" data-line-number="14"></a>
<a class="sourceLine" id="cb5-15" data-line-number="15"><span class="co"># output</span></a>
<a class="sourceLine" id="cb5-16" data-line-number="16">output =<span class="st"> '/neurospin/tmp/ymekki/new_repo/gwhap/T1_subjects/haplotypes'</span></a></code></pre></div>
<p>The function handle one blocs at a time. In order to determine the haplotypes for the blocs created using the whole genome, we need to loop over them.</p>
<p>The determination of the haplotypes for 119.000 blocs is estimated to 1.5 days. In order to reduce the run time, we use the mcmapply function. Basically, this function will paralelize the process.</p>
<p>Please refer to the code below to implement the determination of haplotypes using all the blocs determined in the above step.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="cf">for</span>(chromosome <span class="cf">in</span> <span class="kw">unique</span>(df_blocs<span class="op">$</span>chr)){</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"></a>
<a class="sourceLine" id="cb6-3" data-line-number="3">  <span class="co"># get the start an end bp position of each bloc</span></a>
<a class="sourceLine" id="cb6-4" data-line-number="4">  start     =<span class="st"> </span>df_blocs<span class="op">$</span>fromBp[df_blocs<span class="op">$</span>chr <span class="op">==</span><span class="st"> </span>chromosome]</a>
<a class="sourceLine" id="cb6-5" data-line-number="5">  end       =<span class="st"> </span>df_blocs<span class="op">$</span>toBp[df_blocs<span class="op">$</span>chr <span class="op">==</span><span class="st"> </span>chromosome]</a>
<a class="sourceLine" id="cb6-6" data-line-number="6"></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"></a>
<a class="sourceLine" id="cb6-8" data-line-number="8">  <span class="co">#__________________________________________________________________________________________________</span></a>
<a class="sourceLine" id="cb6-9" data-line-number="9">  <span class="co"># In the bgen file, the IID are annonymized.</span></a>
<a class="sourceLine" id="cb6-10" data-line-number="10">  <span class="co"># It is constitued of, among others, a list of sample that contain the anonymized IID indexed.</span></a>
<a class="sourceLine" id="cb6-11" data-line-number="11">  <span class="co"># The .sample file is used as a reference table. It contains the IID of the participant indexed.</span></a>
<a class="sourceLine" id="cb6-12" data-line-number="12">  <span class="co"># Here, the index can be used as a foreign key to toggle from the anonymized IID to the real ones.</span></a>
<a class="sourceLine" id="cb6-13" data-line-number="13">  <span class="co"># Warning: please notice that .sample file has two header lines.</span></a>
<a class="sourceLine" id="cb6-14" data-line-number="14">  <span class="co"># Be sure to remove the second header and starting the index account from the second line.</span></a>
<a class="sourceLine" id="cb6-15" data-line-number="15">  <span class="co">#__________________________________________________________________________________________________</span></a>
<a class="sourceLine" id="cb6-16" data-line-number="16"></a>
<a class="sourceLine" id="cb6-17" data-line-number="17"></a>
<a class="sourceLine" id="cb6-18" data-line-number="18">  <span class="co"># set bgen chromosome file path</span></a>
<a class="sourceLine" id="cb6-19" data-line-number="19">  bgen_chromosome_file_path =<span class="st"> </span><span class="kw">sprintf</span>(<span class="st">'%s_%s_v2.bgen'</span>, bgen_file_path, chromosome)</a>
<a class="sourceLine" id="cb6-20" data-line-number="20"></a>
<a class="sourceLine" id="cb6-21" data-line-number="21">  <span class="co"># read the index file</span></a>
<a class="sourceLine" id="cb6-22" data-line-number="22">  sample_index_file =<span class="st"> </span><span class="kw">read_delim</span>(<span class="kw">sprintf</span>(<span class="st">'%s_%s_v2_s487395.sample'</span>, sample_file_path, chromosome), <span class="dt">delim=</span><span class="st">'  '</span>)</a>
<a class="sourceLine" id="cb6-23" data-line-number="23"></a>
<a class="sourceLine" id="cb6-24" data-line-number="24">  <span class="co"># remove the first line</span></a>
<a class="sourceLine" id="cb6-25" data-line-number="25">  sample_index_file =<span class="st"> </span>sample_index_file[<span class="op">-</span><span class="dv">1</span>, ]</a>
<a class="sourceLine" id="cb6-26" data-line-number="26"></a>
<a class="sourceLine" id="cb6-27" data-line-number="27">  <span class="co"># get the index as columns</span></a>
<a class="sourceLine" id="cb6-28" data-line-number="28">  <span class="kw">setDT</span>(sample_index_file, <span class="dt">keep.rownames =</span> <span class="ot">TRUE</span>)[]</a>
<a class="sourceLine" id="cb6-29" data-line-number="29"></a>
<a class="sourceLine" id="cb6-30" data-line-number="30"></a>
<a class="sourceLine" id="cb6-31" data-line-number="31">  <span class="co"># filtre on the phenotype IID and get the index as well as the participant IID ordered</span></a>
<a class="sourceLine" id="cb6-32" data-line-number="32">  samples_index =<span class="st"> </span><span class="kw">as.numeric</span>(sample_index_file[sample_index_file<span class="op">$</span>ID_<span class="dv">2</span> <span class="op">%in%</span><span class="st"> </span>phenotype<span class="op">$</span>IID, ]<span class="op">$</span>rn)</a>
<a class="sourceLine" id="cb6-33" data-line-number="33">  samples_index_ID =<span class="st"> </span><span class="kw">as.numeric</span>(sample_index_file[sample_index_file<span class="op">$</span>ID_<span class="dv">2</span> <span class="op">%in%</span><span class="st"> </span>phenotype<span class="op">$</span>IID, ]<span class="op">$</span>ID_<span class="dv">2</span>)</a>
<a class="sourceLine" id="cb6-34" data-line-number="34"></a>
<a class="sourceLine" id="cb6-35" data-line-number="35">  <span class="co"># set bgi file path</span></a>
<a class="sourceLine" id="cb6-36" data-line-number="36">  bgi_file_path =<span class="st"> </span><span class="kw">sprintf</span>(<span class="st">&quot;%s.bgi&quot;</span>, bgen_chromosome_file_path)</a>
<a class="sourceLine" id="cb6-37" data-line-number="37"></a>
<a class="sourceLine" id="cb6-38" data-line-number="38">  <span class="co"># get snps position using the bgi file</span></a>
<a class="sourceLine" id="cb6-39" data-line-number="39">  allVar =<span class="st"> </span><span class="kw">get_bgi_file</span>(<span class="dt">file_path =</span> bgi_file_path)</a>
<a class="sourceLine" id="cb6-40" data-line-number="40"></a>
<a class="sourceLine" id="cb6-41" data-line-number="41">  <span class="co"># get the bgen ID codification</span></a>
<a class="sourceLine" id="cb6-42" data-line-number="42">  dummybg =<span class="st"> </span><span class="kw">get_bgen_file</span>(<span class="dt">file_path =</span> bgen_chromosome_file_path,</a>
<a class="sourceLine" id="cb6-43" data-line-number="43">                          <span class="dt">start =</span> allVar<span class="op">$</span>position[<span class="dv">1</span>],</a>
<a class="sourceLine" id="cb6-44" data-line-number="44">                          <span class="dt">end =</span> allVar<span class="op">$</span>position[<span class="dv">1</span>],</a>
<a class="sourceLine" id="cb6-45" data-line-number="45">                          <span class="dt">samples =</span> <span class="kw">c</span>(),</a>
<a class="sourceLine" id="cb6-46" data-line-number="46">                          <span class="dt">chromosome =</span> <span class="st">''</span>,</a>
<a class="sourceLine" id="cb6-47" data-line-number="47">                          <span class="dt">max_entries_per_sample =</span> <span class="dv">4</span>)</a>
<a class="sourceLine" id="cb6-48" data-line-number="48"></a>
<a class="sourceLine" id="cb6-49" data-line-number="49">  <span class="co"># filter the partcipant bgen ID using their index</span></a>
<a class="sourceLine" id="cb6-50" data-line-number="50">  mysamples =<span class="st"> </span>dummybg<span class="op">$</span>samples[samples_index]</a>
<a class="sourceLine" id="cb6-51" data-line-number="51"></a>
<a class="sourceLine" id="cb6-52" data-line-number="52">  <span class="co"># determine haplotypes for all blocs of one chromosome</span></a>
<a class="sourceLine" id="cb6-53" data-line-number="53">  haplotype_combined =<span class="st"> </span><span class="kw">determine_haplotypes_per_chromosome</span>(<span class="dt">chromosome=</span>chromosome,</a>
<a class="sourceLine" id="cb6-54" data-line-number="54">                                                           start,</a>
<a class="sourceLine" id="cb6-55" data-line-number="55">                                                           end,</a>
<a class="sourceLine" id="cb6-56" data-line-number="56">                                                           <span class="dt">bgen_file =</span> bgen_chromosome_file_path,</a>
<a class="sourceLine" id="cb6-57" data-line-number="57">                                                           <span class="dt">sample_iid =</span> samples_index_ID,</a>
<a class="sourceLine" id="cb6-58" data-line-number="58">                                                           <span class="dt">sample_bgen_iid_code =</span> mysamples,</a>
<a class="sourceLine" id="cb6-59" data-line-number="59">                                                           <span class="dt">max_entries_per_sample=</span><span class="dv">4</span>,</a>
<a class="sourceLine" id="cb6-60" data-line-number="60">                                                           <span class="dt">nb_core=</span><span class="kw">detectCores</span>()<span class="op">-</span><span class="dv">2</span>,</a>
<a class="sourceLine" id="cb6-61" data-line-number="61">                                                           <span class="dt">verbose=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb6-62" data-line-number="62"></a>
<a class="sourceLine" id="cb6-63" data-line-number="63">  <span class="co"># save the haplotypes in tsv file (very heavy, you should consider to use an other way to save the haplotypes)</span></a>
<a class="sourceLine" id="cb6-64" data-line-number="64">  <span class="co">#save_haplotypes(haplotype_combined, chromosome=chr, output=output)</span></a>
<a class="sourceLine" id="cb6-65" data-line-number="65"></a>
<a class="sourceLine" id="cb6-66" data-line-number="66">  <span class="co"># save the haplotypes in RData object</span></a>
<a class="sourceLine" id="cb6-67" data-line-number="67">  <span class="kw">save</span>(haplotype_combined, <span class="dt">file=</span><span class="kw">sprintf</span>(<span class="st">'%s/haplotypes_%s.RData'</span>, output, chromosome), <span class="dt">compress=</span>T)</a>
<a class="sourceLine" id="cb6-68" data-line-number="68">}</a></code></pre></div>
<p>At the end, the haplotypes concerning one chromosome are binded by column and then saved.</p>
<p>Since, the test step implies that the haplotypes are tested per blocs, you need to reverse the binding process. Please refer to the section below to get how to do that.</p>
</div>
<div id="test-the-haplotypes" class="section level1">
<h1>Test the haplotypes</h1>
<p>The function is designed to handle one blocs at the time. In order to apply the three test on all haplotypes, we need to loop over them.</p>
<p>In order to reduce the run time, we use the mcmapply function. Basically, this function will paralelize the process.</p>
<p>Please refer to the code below to implement the tests using all haplotypes determined in the above step.</p>
<ul>
<li>Set parameters</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">chr =<span class="st"> &quot;chr1&quot;</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2"></a>
<a class="sourceLine" id="cb7-3" data-line-number="3"><span class="co"># filtre sur les IDs common phenotype and genotype</span></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="co"># read index file</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5">sample_semi_path =<span class="st"> '/neurospin/ukb/genetic/GENETIC_DATA_500k/HAPLOTYPES/v2/ukb25251_hap'</span></a>
<a class="sourceLine" id="cb7-6" data-line-number="6">sample_index_file =<span class="st"> </span><span class="kw">read_delim</span>(<span class="kw">sprintf</span>(<span class="st">'%s_%s_v2_s487395.sample'</span>, sample_semi_path, chr), <span class="dt">delim=</span><span class="st">'  '</span>)</a>
<a class="sourceLine" id="cb7-7" data-line-number="7">sample_index_file =<span class="st"> </span>sample_index_file[<span class="op">-</span><span class="dv">1</span>, ]</a>
<a class="sourceLine" id="cb7-8" data-line-number="8"></a>
<a class="sourceLine" id="cb7-9" data-line-number="9">phenotypes_sulci =<span class="st"> </span><span class="kw">fread</span>(<span class="st">&quot;/neurospin/ukb/workspace/pheno_boxcox/all_pheno_res.tsv&quot;</span>)</a>
<a class="sourceLine" id="cb7-10" data-line-number="10">Y_filtred =<span class="st"> </span>phenotypes_sulci[, <span class="kw">c</span>(<span class="st">'IID'</span>, <span class="st">'FCMpost_left'</span>)]</a>
<a class="sourceLine" id="cb7-11" data-line-number="11"></a>
<a class="sourceLine" id="cb7-12" data-line-number="12">output =<span class="st"> '/neurospin/tmp/ymekki/new_repo/gwhap/T1_subjects/lm_test'</span></a></code></pre></div>
<ul>
<li>Apply the test</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">22</span>){</a>
<a class="sourceLine" id="cb8-2" data-line-number="2"></a>
<a class="sourceLine" id="cb8-3" data-line-number="3">  output_chromosome =<span class="st"> </span><span class="kw">sprintf</span>(<span class="st">'%s/chr_%s'</span>, output, i)</a>
<a class="sourceLine" id="cb8-4" data-line-number="4"></a>
<a class="sourceLine" id="cb8-5" data-line-number="5">  <span class="kw">dir.create</span>(output_chromosome)</a>
<a class="sourceLine" id="cb8-6" data-line-number="6">  <span class="kw">dir.create</span>(<span class="kw">sprintf</span>(<span class="st">'%s/bloc_test_results'</span>, output_chromosome))</a>
<a class="sourceLine" id="cb8-7" data-line-number="7">  <span class="kw">dir.create</span>(<span class="kw">sprintf</span>(<span class="st">'%s/complete_test_results'</span>, output_chromosome))</a>
<a class="sourceLine" id="cb8-8" data-line-number="8">  <span class="kw">dir.create</span>(<span class="kw">sprintf</span>(<span class="st">'%s/single_test_results'</span>, output_chromosome))</a>
<a class="sourceLine" id="cb8-9" data-line-number="9">  </a>
<a class="sourceLine" id="cb8-10" data-line-number="10">  </a>
<a class="sourceLine" id="cb8-11" data-line-number="11">  <span class="co"># load haplotype_combined</span></a>
<a class="sourceLine" id="cb8-12" data-line-number="12">  haplotypes_path =<span class="st"> '/neurospin/tmp/ymekki/new_repo/gwhap/T1_subjects/haplotypes/old_haplotypes_'</span></a>
<a class="sourceLine" id="cb8-13" data-line-number="13">  <span class="kw">load</span>(<span class="kw">sprintf</span>(<span class="st">'%schr%s.RData'</span>, haplotypes_path, i))</a>
<a class="sourceLine" id="cb8-14" data-line-number="14">  </a>
<a class="sourceLine" id="cb8-15" data-line-number="15">  </a>
<a class="sourceLine" id="cb8-16" data-line-number="16">  chr =<span class="st"> </span><span class="kw">sprintf</span>(<span class="st">'chr%s'</span>, i)</a>
<a class="sourceLine" id="cb8-17" data-line-number="17"></a>
<a class="sourceLine" id="cb8-18" data-line-number="18">  <span class="co">#_____________________________________________________________________________________</span></a>
<a class="sourceLine" id="cb8-19" data-line-number="19">  <span class="co"># During the haplotypes determination steps,</span></a>
<a class="sourceLine" id="cb8-20" data-line-number="20">  <span class="co"># the haplotypes concerning one chromosome were binded by column and then saved.</span></a>
<a class="sourceLine" id="cb8-21" data-line-number="21">  <span class="co"># The test haplotype function implies that each bloc is testes separtelty.</span></a>
<a class="sourceLine" id="cb8-22" data-line-number="22">  <span class="co"># Therefore, you need to reverse the binding process. This is done as following:</span></a>
<a class="sourceLine" id="cb8-23" data-line-number="23">  <span class="co"># first, identify the blocs using the haplotypes code (see code below)</span></a>
<a class="sourceLine" id="cb8-24" data-line-number="24">  <span class="co"># then, filter the haplotypes object using the bloc code identified above</span></a>
<a class="sourceLine" id="cb8-25" data-line-number="25">  <span class="co"># (see lm_test_haplotypes_per_bloc function)</span></a>
<a class="sourceLine" id="cb8-26" data-line-number="26">  <span class="co">#_____________________________________________________________________________________</span></a>
<a class="sourceLine" id="cb8-27" data-line-number="27">  </a>
<a class="sourceLine" id="cb8-28" data-line-number="28">  <span class="co"># separate the blocs</span></a>
<a class="sourceLine" id="cb8-29" data-line-number="29">  </a>
<a class="sourceLine" id="cb8-30" data-line-number="30">  <span class="co"># replace NA added by cbind by the chromosome code</span></a>
<a class="sourceLine" id="cb8-31" data-line-number="31">  <span class="kw">colnames</span>(haplotype_combined) =<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;NA&quot;</span>, chr, <span class="kw">colnames</span>(haplotype_combined))</a>
<a class="sourceLine" id="cb8-32" data-line-number="32">  </a>
<a class="sourceLine" id="cb8-33" data-line-number="33">  <span class="co"># get the start and end bloc's position</span></a>
<a class="sourceLine" id="cb8-34" data-line-number="34">  start =<span class="st"> </span><span class="kw">na.omit</span>(<span class="kw">unique</span>(<span class="kw">vapply</span>(<span class="kw">strsplit</span>(<span class="kw">colnames</span>(haplotype_combined),<span class="st">&quot;_&quot;</span>), <span class="st">`</span><span class="dt">[</span><span class="st">`</span>, <span class="dv">2</span>, <span class="dt">FUN.VALUE=</span><span class="kw">character</span>(<span class="dv">1</span>))))</a>
<a class="sourceLine" id="cb8-35" data-line-number="35">  end =<span class="st"> </span><span class="kw">na.omit</span>(<span class="kw">unique</span>(<span class="kw">vapply</span>(<span class="kw">strsplit</span>(<span class="kw">colnames</span>(haplotype_combined),<span class="st">&quot;_&quot;</span>), <span class="st">`</span><span class="dt">[</span><span class="st">`</span>, <span class="dv">3</span>, <span class="dt">FUN.VALUE=</span><span class="kw">character</span>(<span class="dv">1</span>))))</a>
<a class="sourceLine" id="cb8-36" data-line-number="36">  </a>
<a class="sourceLine" id="cb8-37" data-line-number="37">  <span class="co"># concatenate the start and end bloc position</span></a>
<a class="sourceLine" id="cb8-38" data-line-number="38">  blocs =<span class="st"> </span><span class="kw">sprintf</span>(<span class="st">'%s_%s'</span>, start, end)</a>
<a class="sourceLine" id="cb8-39" data-line-number="39">  </a>
<a class="sourceLine" id="cb8-40" data-line-number="40">  <span class="co"># run the test</span></a>
<a class="sourceLine" id="cb8-41" data-line-number="41">  results_chr =<span class="st"> </span><span class="kw">mcmapply</span>(<span class="dt">FUN =</span> lm_test_haplotypes_per_bloc,</a>
<a class="sourceLine" id="cb8-42" data-line-number="42">                         <span class="dt">blocs =</span> blocs,</a>
<a class="sourceLine" id="cb8-43" data-line-number="43">                         <span class="dt">haplotype_combined =</span> <span class="kw">list</span>(haplotype_combined),</a>
<a class="sourceLine" id="cb8-44" data-line-number="44">                         <span class="dt">sample_index_file =</span> <span class="kw">list</span>(sample_index_file),</a>
<a class="sourceLine" id="cb8-45" data-line-number="45">                         <span class="dt">Y_filtred =</span> <span class="kw">list</span>(Y_filtred),</a>
<a class="sourceLine" id="cb8-46" data-line-number="46">                         <span class="dt">output_chromosome =</span> output_chromosome,</a>
<a class="sourceLine" id="cb8-47" data-line-number="47">                         <span class="dt">mc.cores =</span> <span class="dv">30</span>)</a>
<a class="sourceLine" id="cb8-48" data-line-number="48">}</a></code></pre></div>
<ul>
<li>Get the summary of the results. It consist on filtering on the signicant p values.</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="kw">summary_haplotypes_test</span>(output, <span class="dt">threshold =</span> <span class="fl">5e-6</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>)</a></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">lm_test_haplotypes_per_bloc =<span class="st"> </span><span class="cf">function</span>(blocs, haplotype_combined, sample_index_file, Y_filtred, output_chromosome, <span class="dt">kind=</span><span class="st">'all'</span>){</a>
<a class="sourceLine" id="cb10-2" data-line-number="2"></a>
<a class="sourceLine" id="cb10-3" data-line-number="3">  <span class="co"># change Y_filtred column name</span></a>
<a class="sourceLine" id="cb10-4" data-line-number="4">  <span class="kw">colnames</span>(Y_filtred) =<span class="st"> </span><span class="kw">c</span>(<span class="st">'IID'</span>, <span class="st">'phenotype'</span>)</a>
<a class="sourceLine" id="cb10-5" data-line-number="5"></a>
<a class="sourceLine" id="cb10-6" data-line-number="6">  <span class="co"># get only the haplotypes corresponding to bloc got as input</span></a>
<a class="sourceLine" id="cb10-7" data-line-number="7">  columns_blocs =<span class="st"> </span><span class="kw">str_subset</span>(<span class="kw">colnames</span>(haplotype_combined), blocs)</a>
<a class="sourceLine" id="cb10-8" data-line-number="8">  </a>
<a class="sourceLine" id="cb10-9" data-line-number="9">  <span class="co"># filter the haplotypes corresponding to the bloc</span></a>
<a class="sourceLine" id="cb10-10" data-line-number="10">  haplotype_one_bloc =<span class="st"> </span>haplotype_combined[, columns_blocs]</a>
<a class="sourceLine" id="cb10-11" data-line-number="11"></a>
<a class="sourceLine" id="cb10-12" data-line-number="12">  <span class="co"># This is done in order to be sure that the subject are well alligned</span></a>
<a class="sourceLine" id="cb10-13" data-line-number="13">  </a>
<a class="sourceLine" id="cb10-14" data-line-number="14">  <span class="co"># get the index columns</span></a>
<a class="sourceLine" id="cb10-15" data-line-number="15">  <span class="kw">setDT</span>(haplotype_one_bloc, <span class="dt">keep.rownames =</span> <span class="ot">TRUE</span>)[]</a>
<a class="sourceLine" id="cb10-16" data-line-number="16"></a>
<a class="sourceLine" id="cb10-17" data-line-number="17">  <span class="co"># merge the index file and the haplotypes using the index</span></a>
<a class="sourceLine" id="cb10-18" data-line-number="18">  haplotype_one_bloc<span class="op">$</span>rn &lt;-<span class="st"> </span><span class="kw">as.integer</span>(haplotype_one_bloc<span class="op">$</span>rn)</a>
<a class="sourceLine" id="cb10-19" data-line-number="19">  df_tmp_merged =<span class="st"> </span><span class="kw">merge</span>(sample_index_file, haplotype_one_bloc, <span class="dt">by.x=</span><span class="st">'ID_2'</span>, <span class="dt">by.y=</span><span class="st">&quot;rn&quot;</span>)</a>
<a class="sourceLine" id="cb10-20" data-line-number="20"></a>
<a class="sourceLine" id="cb10-21" data-line-number="21">  <span class="co"># merge the df obtained above with the phenotype using the ID</span></a>
<a class="sourceLine" id="cb10-22" data-line-number="22">  df_merged =<span class="st"> </span><span class="kw">merge</span>(df_tmp_merged, Y_filtred, <span class="dt">by.x =</span> <span class="st">'ID_2'</span>, <span class="dt">by.y =</span> <span class="st">'IID'</span>)</a>
<a class="sourceLine" id="cb10-23" data-line-number="23"></a>
<a class="sourceLine" id="cb10-24" data-line-number="24">  <span class="co"># set X, Y and run the haplotypique test</span></a>
<a class="sourceLine" id="cb10-25" data-line-number="25">  X =<span class="st"> </span>df_merged[, columns_blocs]</a>
<a class="sourceLine" id="cb10-26" data-line-number="26">  Y =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">phenotype =</span> df_merged[, <span class="st">'phenotype'</span>])</a>
<a class="sourceLine" id="cb10-27" data-line-number="27">  </a>
<a class="sourceLine" id="cb10-28" data-line-number="28">  <span class="co"># perform the test</span></a>
<a class="sourceLine" id="cb10-29" data-line-number="29">  results =<span class="st"> </span><span class="kw">lm_test_haplotypes</span>(X, Y, <span class="dt">kind=</span>kind)</a>
<a class="sourceLine" id="cb10-30" data-line-number="30"></a>
<a class="sourceLine" id="cb10-31" data-line-number="31">  <span class="co"># save the results</span></a>
<a class="sourceLine" id="cb10-32" data-line-number="32">  <span class="kw">write.table</span>(<span class="kw">data.frame</span>(results<span class="op">$</span>bloc), <span class="kw">sprintf</span>(<span class="st">'%s/bloc_test_results/%s.tsv'</span>, output_chromosome, blocs), <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names=</span><span class="ot">FALSE</span>, <span class="dt">quote=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb10-33" data-line-number="33">  <span class="kw">write.table</span>(<span class="kw">data.frame</span>(results<span class="op">$</span>complete), <span class="kw">sprintf</span>(<span class="st">'%s/complete_test_results/%s.tsv'</span>, output_chromosome, blocs), <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names=</span><span class="ot">FALSE</span>, <span class="dt">quote=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb10-34" data-line-number="34">  <span class="kw">write.table</span>(<span class="kw">data.frame</span>(results<span class="op">$</span>single), <span class="kw">sprintf</span>(<span class="st">'%s/single_test_results/%s.tsv'</span>, output_chromosome, blocs), <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names=</span><span class="ot">FALSE</span>, <span class="dt">quote=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb10-35" data-line-number="35"></a>
<a class="sourceLine" id="cb10-36" data-line-number="36">  <span class="kw">return</span>(results)</a>
<a class="sourceLine" id="cb10-37" data-line-number="37">}</a></code></pre></div>
<ul>
<li>Residualisation</li>
</ul>
<p>If you need to residualise your phenotypes, please refer to this code as an example:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">covariates =<span class="st"> </span><span class="kw">fread</span>(<span class="st">'/neurospin/ukb/workspace/pheno_boxcox/covar_SexAgeMRI_10PCSArray_19k.cov'</span>)</a>
<a class="sourceLine" id="cb11-2" data-line-number="2">X_covariates_names =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;IID&quot;</span>, <span class="st">&quot;Array&quot;</span>, <span class="st">&quot;Sex&quot;</span>, <span class="st">&quot;Age&quot;</span>, <span class="st">&quot;PC1&quot;</span>, <span class="st">&quot;PC2&quot;</span>, <span class="st">&quot;PC3&quot;</span>, <span class="st">&quot;PC4&quot;</span>, <span class="st">&quot;PC5&quot;</span>, <span class="st">&quot;PC6&quot;</span>, <span class="st">&quot;PC7&quot;</span>, <span class="st">&quot;PC8&quot;</span>, <span class="st">&quot;PC9&quot;</span>, <span class="st">&quot;PC10&quot;</span>)</a>
<a class="sourceLine" id="cb11-3" data-line-number="3"></a>
<a class="sourceLine" id="cb11-4" data-line-number="4">phenotypes_sulci =<span class="st"> </span><span class="kw">fread</span>(<span class="st">&quot;/neurospin/ukb/workspace/pheno_boxcox/all_sulci_boxcox.tsv&quot;</span>)</a>
<a class="sourceLine" id="cb11-5" data-line-number="5">phenotype_FCMpost_left =<span class="st"> </span>phenotypes_sulci[, <span class="kw">c</span>(<span class="st">'IID'</span>, <span class="st">'FCMpost_left'</span>)]</a>
<a class="sourceLine" id="cb11-6" data-line-number="6"></a>
<a class="sourceLine" id="cb11-7" data-line-number="7">Y_residualised =<span class="st"> </span><span class="kw">phenotype_residualisation</span>(<span class="dt">phenotype=</span>phenotype_FCMpost_left, <span class="dt">covariates=</span>covariates[, ..X_covariates_names], <span class="dt">verbose=</span><span class="ot">FALSE</span>)</a></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
