<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>How to use gwhap package</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">How to use gwhap package</h1>


<div id="TOC">
<ul>
<li><a href="#installation-of-gwhap">Installation of <code>gwhap</code></a></li>
<li><a href="#create-the-augmented-genetic-map">Create the augmented genetic map</a></li>
<li><a href="#blocs-creation">Blocs creation</a><ul>
<li><a href="#create-the-blocs">Create the blocs</a></li>
<li><a href="#blocs-visualization">Blocs visualization</a></li>
</ul></li>
<li><a href="#determine-the-haplotypes">Determine the haplotypes</a></li>
<li><a href="#test-the-haplotypes">Test the haplotypes</a></li>
</ul>
</div>

<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(gwhap)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"></a>
<a class="sourceLine" id="cb1-3" data-line-number="3">RESOURCE_ROOT =<span class="st"> &quot;&lt;YOUR_LOCAL_RESOURCE_FS&gt;&quot;</span></a></code></pre></div>
<div id="installation-of-gwhap" class="section level1">
<h1>Installation of <code>gwhap</code></h1>
<p>There are two ways to obtain the gwhap source package:</p>
<ul>
<li>Using the CRAN</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">install.packages</span>(<span class="st">'gwhap'</span>)</a></code></pre></div>
<ul>
<li>Using github</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">'yasmina-mekki/gwhap'</span>)</a></code></pre></div>
<p>The gwhap package implies the use of a version of R <strong>(3.6.2 and higher)</strong> as well as a certain number libraries.</p>
<p>In order to facilitate the use of it, here is a recipe of a singularity image containing all the necessary. Please refer to the github repo, for more explanation about how to generate it.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;yasmina-mekki/gwhap_singularity_image&quot;</span>)</a></code></pre></div>
</div>
<div id="create-the-augmented-genetic-map" class="section level1">
<h1>Create the augmented genetic map</h1>
<p>Currently, only threee kind of map are took into account:</p>
<ul>
<li>Rutgers genetic map v3 (Kosambi)</li>
<li>1000_genome</li>
<li>1000_genome_interpolated</li>
</ul>
<p>Here is an example using the Rutgers genetic map.</p>
<p>Download the Rutgers genetic map v3 (Kosambi)</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">download_rutgers_map</span>(RESOURCE_ROOT)</a></code></pre></div>
<p>interpolate a new map for SNPs using either:</p>
<ul>
<li>bim file</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">snp_physical_positions =<span class="st"> &quot;path_to_your_snps_file/snp_file.bim&quot;</span></a></code></pre></div>
<ul>
<li>bgi/bgen file</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">snp_physical_positions =<span class="st"> &quot;path_to_your_bgen_file/ukb_hap_&quot;</span></a></code></pre></div>
<p>Create the augmented genetic map as follow:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw">create_augmented_genetic_map</span>(<span class="dt">snp_physical_positions=</span>snp_physical_positions, <span class="dt">genetic_map_dir=</span><span class="st">'path_to_your_rutgers_map_dir'</span>, <span class="dt">map_name=</span><span class="st">'rutgers'</span>, <span class="dt">save_genetic_map=</span><span class="ot">TRUE</span>, <span class="dt">output=</span><span class="st">'path_to_your_output_dir'</span>)</a></code></pre></div>
</div>
<div id="blocs-creation" class="section level1">
<h1>Blocs creation</h1>
<div id="create-the-blocs" class="section level2">
<h2>Create the blocs</h2>
<p>Before creating the clocs, you need to load the map created above</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">augmented_map_df =<span class="st"> </span><span class="kw">get_augmented_genetic_map</span>(<span class="dt">augmented_genetic_map_dir=</span><span class="st">'path_to_your_output_dir'</span>, <span class="dt">chromosomes=</span><span class="dv">1</span><span class="op">:</span><span class="dv">23</span>)</a></code></pre></div>
<p>Create blocs with a single delta</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="kw">create_blocs</span>(<span class="dt">genetics_map=</span>augmented_map_df, <span class="dt">delta=</span><span class="fl">1e-3</span>, <span class="dt">save_blocs=</span><span class="ot">TRUE</span>, <span class="dt">output=</span><span class="st">'path_to_your_output_dir'</span>)</a></code></pre></div>
<p>Create blocs with multiple values of delta</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">deltas =<span class="st"> </span><span class="kw">c</span>(<span class="fl">1e-3</span>, <span class="fl">2.5e-3</span>, <span class="fl">5e-3</span>, <span class="fl">7.5e-3</span>)</a>
<a class="sourceLine" id="cb11-2" data-line-number="2">dfs_blocks =<span class="st"> </span><span class="kw">data.frame</span>()</a>
<a class="sourceLine" id="cb11-3" data-line-number="3"><span class="cf">for</span>(delta <span class="cf">in</span> deltas){</a>
<a class="sourceLine" id="cb11-4" data-line-number="4">  df_blocks_tmp =<span class="st"> </span><span class="kw">create_blocs</span>(<span class="dt">genetics_map=</span>augmented_map_df, <span class="dt">delta=</span>delta, <span class="dt">save_blocs=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb11-5" data-line-number="5">  dfs_blocks &lt;-<span class="st"> </span><span class="kw">rbind</span>(dfs_blocks, df_blocks_tmp)</a>
<a class="sourceLine" id="cb11-6" data-line-number="6">}</a></code></pre></div>
</div>
<div id="blocs-visualization" class="section level2">
<h2>Blocs visualization</h2>
<p>Assuming that df_blocs is your bloc data frame already loaded</p>
<p>Blocs distribution plot</p>
<ul>
<li>Blocs distribution for all chromosomes and optionnaly for different values of delta</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">haplotype_bloc_distribution_per_delta =<span class="st"> </span><span class="kw">haplotype_bloc_distribution</span>(df_blocs)</a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="kw">save_plot</span>(haplotype_bloc_distribution_per_delta, <span class="kw">paste0</span>(<span class="st">'output_path/'</span>, <span class="st">'haplotype_bloc_distribution_per_delta.png'</span>))</a></code></pre></div>
<ul>
<li>Blocs distribution per chromosome and optionnaly for different values of delta</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">haplotype_bloc_distribution_per_chr =<span class="st"> </span><span class="kw">haplotype_bloc_distribution</span>(df_blocs, <span class="dt">per_chromosome=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb13-2" data-line-number="2"><span class="kw">save_plot</span>(haplotype_bloc_distribution_per_chr, <span class="kw">paste0</span>(<span class="st">'output_path/'</span>, <span class="st">'haplotype_bloc_distribution_per_chr.png'</span>))</a></code></pre></div>
<p>Karyotype plot</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">karyotype_plot_obj =<span class="st"> </span><span class="kw">karyotype_plot</span>(df_blocs)</a>
<a class="sourceLine" id="cb14-2" data-line-number="2"><span class="kw">save_plot</span>(karyotype_plot_obj, <span class="kw">paste0</span>(<span class="st">'output_path/'</span>, <span class="st">'karyotype_plot.png'</span>))</a></code></pre></div>
</div>
</div>
<div id="determine-the-haplotypes" class="section level1">
<h1>Determine the haplotypes</h1>
<p>Before determining the haplotypes, you need to load the blocs created above</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">blocs_df =<span class="st"> </span><span class="kw">get_blocs</span>(<span class="dt">augmented_genetic_map_dir=</span><span class="st">'path_to_your_output_dir'</span>)</a></code></pre></div>
<p>Here is an example for one bloc</p>
<ul>
<li>Set the parameters</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">chr =<span class="st"> 'chr1'</span></a>
<a class="sourceLine" id="cb16-2" data-line-number="2">start =<span class="st"> </span><span class="dv">1000</span></a>
<a class="sourceLine" id="cb16-3" data-line-number="3">end =<span class="st"> </span><span class="dv">1200</span></a>
<a class="sourceLine" id="cb16-4" data-line-number="4">bgnfile =<span class="st"> </span><span class="kw">sprintf</span>(<span class="st">'bgen_file_path_%s.bgen'</span>, chr)</a>
<a class="sourceLine" id="cb16-5" data-line-number="5">samples_index =<span class="st"> </span><span class="kw">c</span>(<span class="st">'index_1'</span>, <span class="st">'index_2'</span>, <span class="st">'index_3'</span>, <span class="st">'index_4'</span>, <span class="st">'index_5'</span>)</a>
<a class="sourceLine" id="cb16-6" data-line-number="6">mysamples =<span class="st"> </span><span class="kw">c</span>(<span class="st">'IID_1'</span>, <span class="st">'IID_2'</span>, <span class="st">'IID_3'</span>, <span class="st">'IID_4'</span>, <span class="st">'IID_5'</span>)</a></code></pre></div>
<ul>
<li>Run the determine haplotype function</li>
</ul>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">haplotypes =<span class="st"> </span><span class="kw">determine_haplotypes_per_bloc</span>(chr, start, end, bgnfile, samples_index, mysamples)</a></code></pre></div>
<p>Please refer to the genome wide example vignette for more code about how to determine the haplotypes using all the blocs.</p>
<p><strong><em>Note: </em></strong></p>
<p>In the bgen file, the IID are annonymized. It is constitued of, among others, a list of sample that contain the anonymized IID indexed. The .sample file is used as a reference table. It contains the IID of the participant indexed. Here, the index can be used as a foreign key to toggle from the anonymized IID to the real ones.</p>
<p><strong><em>Warning:</em></strong> please notice that <a href="https://www.cog-genomics.org/plink/2.0/formats#sample">.sample</a> file has two header lines. Be sure to remove the second header and starting the index account from the second line.</p>
</div>
<div id="test-the-haplotypes" class="section level1">
<h1>Test the haplotypes</h1>
<ul>
<li>Set the parameters</li>
</ul>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">X =<span class="st"> </span><span class="kw">load</span>(<span class="st">'haplotypes_path_chr1.RData'</span>)</a>
<a class="sourceLine" id="cb18-2" data-line-number="2">Y =<span class="st"> </span><span class="kw">read_delim</span>(file_path_phenotype.tsv, <span class="dt">delim=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</a></code></pre></div>
<ul>
<li>Run the test haplotypes function</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">test_results =<span class="st"> </span><span class="kw">lm_test_haplotypes</span>(X, Y, <span class="dt">kind=</span><span class="st">'all'</span>)</a></code></pre></div>
<p><strong><em>Warning:</em></strong></p>
<p>One should know that the test haplotype function assume that the Y are residualized. Indeed, it does not take into account the covariates.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
