---
title: "How to use gwhap package"
output:
  #rmarkdown::html_vignette:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{how_to_use_gwhap}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(gwhap)

RESOURCE_ROOT = "<YOUR_LOCAL_RESOURCE_FS>"
```


# Installation of `gwhap`

```R
if (!require(devtools)){install.packages("devtools")}
devtools::install_github("yasmina-mekki/gwhap")
```


# Create the augmented genetic map


### Using SNP list and genetic map on the same chromosome

```{r}
snp_list=c(100,150,200)
genetic_map=data.frame(bp=c(100,200),
                        cM=c(0.0,0.01))
new_map=genetic_map_interp_chr(genetic_map = genetic_map,
                              snp_list = snp_list)
head(new_map)

```
### Using SNP list and genetic map on several chromosome
We have 2 regions of 3 SNPs in 2 chromosomes chr1  and chr2 : 
```{r}
snp_list=data.frame(chr=c(rep("chr1",3),rep("chr2",3)),
                    bp=c(100,150,200,100,150,200))
head(snp_list)
```
We create a reference genetic map :
```{r}

genetic_map=data.frame(chr=c(rep("chr1",2),rep("chr2",2)),
                       bp=c(100,200,100,200),
                        cM=c(0.0,0.01,0.0,0.1))
head(genetic_map)
```
Now we can interpolate the missing posistion in cM for all the chromosomes : 
```{r}
new_map=genetic_map_interp(genetic_map_all  = genetic_map,
                                snp_list  = snp_list)
head(new_map)

```
###Using snp file and Rutgers Map


Download the Rutgers genetic map v3 (Kosambi)
```R
download_rutgers_map(RESOURCE_ROOT)
```

interpolate a new map for SNPs in snp_file.txt
```R
to_be_updated_genetic_map_path = "path_to_your_snps_file/snp_file.txt"
rutgers_map = "path_to_your_rutgers_map"
gen_map_updated = read.augmented_genetic_map(genetic_map_path=rutgers_map,
                                            to_be_updated_genetic_map_path=to_be_updated_genetic_map_path)
```

# Blocs creation

## Create the blocs

Create blocs with a single delta
```R
df_blocks = create_blocks(gen_map_updated, delta=1e-3)
```

Create blocs with multiple values of delta
```R
deltas = c(1e-3, 2.5e-3, 5e-3, 7.5e-3)
dfs_blocks = data.frame()
for(delta in deltas){
  df_blocks_tmp = create_blocks(genetics_map=gen_map_updated, delta=delta)
  dfs_blocks <- rbind(dfs_blocks, df_blocks_tmp)
}
```

##  Blocs visualization

Blocs distribution plot

* All chromosomes
```
haplotype_block_distribution_per_delta = haplotype_block_distribution(dfs_blocks, xlim=c(1.590790e-10, 1), ylim=c(0, 1.5))
print(haplotype_block_distribution_per_delta)
save_plot(haplotype_block_distribution_per_delta, paste0(output, 'haplotype_block_distribution_per_delta.pdf'))
```

* Per chromosome
```R
haplotype_block_distribution_per_chr = haplotype_block_distribution(dfs_blocks, xlim=c(1.590790e-10, 1), ylim=c(0, 1.5), per_chromosome=TRUE)
print(haplotype_block_distribution_per_chr)
save_plot(haplotype_block_distribution_per_chr, paste0(output, 'haplotype_block_distribution_per_chr.pdf'))
```

Karyotype
```R
df_blocs_delta_3 = dfs_blocks[dfs_blocks$delta==0.001,]
karyotype_plot_obj = karyotype_plot(df_blocs_delta_3)
save_plot(karyotype_plot_obj, paste0(output, 'karyotype_plot.pdf'))
```


# Determine the haplotypes




# Phenotypes distribution plots

Phenotypes distribution plots

```R
phenotypes_path = Sys.glob(file.path("path_to_your_phenotypes/" , '*'))
phenotype_distribution_plot = phenotype_distribution(phenotypes_path = phenotypes_path)
print(phenotype_distribution_plot)
save_plot(phenotype_distribution_plot, paste0(output, 'phenotype_distribution_plot.pdf'))
```



