---
title: "GWHAP PART1 - GENETIC MAPS INTERPOLATION"
author: "SK"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{GWHAP PART1 - GENETIC MAPS INTERPOLATION}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = FALSE,
  error=FALSE,
  message=FALSE,
  warning=FALSE,
  comment = "#>"
)
```
```{r}
library(gwhap)
library(data.table)
library(readr)
```
# Genome-Wide HAPloptype analysis  
## PART 1 : Genetic Map
### GWHAP needs to depermine the location the genetic map (i.e in cM) of the phased markers. 

For every marker postion in bp, GWHAP will try find it in a reference genetic map.
If the marker's position is not in the genetic map, its position in cM is computed using linear interpolation. 
GWHAP can use several reference maps format :  Rutgers genetic map v3 , 1000 Genomes ...
See vignette "Ressources" for details
TO DO : code that check the download URL for all downloaders

### Reading the the reference genetic map :


* GWHAP can read compressed (.gz) 1000 genomes maps using 3 columns [ rsid ; position ; cM]  :
```{r}
# give the list of files to consider
f1 = system.file("extdata", "chr1.interpolated_genetic_map.gz",
                          package="gwhap", mustWork=TRUE)
f2 = system.file("extdata", "chr2.interpolated_genetic_map.gz",
                          package="gwhap", mustWork=TRUE)
# build a NAMED list mapping the chrom information
chr = list(1, 2)
names(chr) = c(f1, f2)

# now build the parameter for Genetic_Map object. Inspect the file to
# assign the right column name for cM and position.
# Here the chrom info is not read from the file it is read from the
# NAMED list chr
filepaths = c(f1, f2)
encodings = list("cM"="cM", "position"="bp","chr"=chr)

# get an instance of Genetic_Map
genetic_map = Genetic_Map(filepaths=filepaths, encodings=encodings)

# finally read the corresponding data and have a quick look at it
genetic_map = readData(genetic_map)
print(head(genetic_map@gmapData))
```

* GWHAP has a helper variable gwhapConfig with the main description of the toy genetic map of the package
```{r}
# Inspect gwhapConfig global varaible and see the descriptions of 
# the various toy genetic map  
names(gwhap:::gwhapConfig)

# get the filepaths and encodings for 1000 genome interpolated
filepaths = gwhap:::gwhapConfig$genmap_toy_interpolated_1000$filepaths
encodings = gwhap:::gwhapConfig$genmap_toy_interpolated_1000$encodings

genetic_map = Genetic_Map(filepaths=filepaths, encodings=encodings)
genetic_map = readData(genetic_map)
print(head(genetic_map@gmapData))
```



* GWHAP can read reference maps in Rutgers Maps filename and column format :
```{r}
# init filepaths and encodings parameters
filepaths = gwhap:::gwhapConfig$genmap_toy_rutger$filepaths
encodings = gwhap:::gwhapConfig$genmap_toy_rutger$encodings


# get an instance of Genetic_Map
genetic_map = Genetic_Map(filepaths=filepaths, encodings=encodings)

# finally read the corresponding data and have a quick look at it
genetic_map = readData(genetic_map)
print(head(genetic_map@gmapData))
```


GWHAP can read 1000 genomes maps:

* GWHAP can read  1000 genomes genetic maps file format
```{r}
# init filepaths and encodings parameters
filepaths = gwhap:::gwhapConfig$genmap_toy_reference_1000$filepaths
encodings = gwhap:::gwhapConfig$genmap_toy_reference_1000$encodings

# get an instance of Genetic_Map
genetic_map = Genetic_Map(filepaths=filepaths, encodings=encodings)

# finally read the corresponding data and have a quick look at it
genetic_map = readData(genetic_map)
print(head(genetic_map@gmapData))
```




### Reading the list of SNPs to place on the map :

* GWHAP can read a list of SNPs using the bgi/bgen file :
```{r}
snp_file=system.file("extdata","haplotypes.bgen.bgi", package="gwhap", mustWork=TRUE)
snp_list=get_bgi_file(snp_file)
head(snp_list)
```

* GWHAP can read a list of SNPs using any text file in .bim format  :
```{r}
snp_physical_positions = system.file("extdata", "example.bim", package="gwhap", mustWork=TRUE)
snp_list=get_bim_file(snp_physical_positions)

head(snp_list)
```


### GWHAP provide build-in function to interpolate a list of SNPs using a reference map :
TODO fix buggy "interp_in" list : keep snps reference needed


*  Augmented maps for a list of SNP read from a bim files format and 1000 Genomes interpolated maps:
```{r}
filepaths = gwhap:::gwhapConfig$genmap_toy_interpolated_1000$filepaths
encodings = gwhap:::gwhapConfig$genmap_toy_interpolated_1000$encodings


# get an instance of Genetic_Map
interp1000_genetic_map = Genetic_Map(filepaths=filepaths, encodings=encodings)
interp1000_genetic_map = readData(interp1000_genetic_map)

# get the snp list position filename
snp_physical_positions = system.file("extdata", "example.bim", package="gwhap", mustWork=TRUE)

# get the augmented genetic_map
interpolated_map=create_augmented_genetic_map(snp_physical_positions=snp_physical_positions, genetic_map=interp1000_genetic_map, genetic_map_dir=genetic_map_dir, map_name='1000_genome_interpolated', save_genetic_map=FALSE)
head(interpolated_map)
```


TODO : Use an ybgen.bgi filenames format (use list?) and  1000 genomes maps

* Using UKB Biobank filenames format (bgen.bgi) and  1000 genomes maps : 

```{r}
filepaths = gwhap:::gwhapConfig$genmap_toy_reference_1000$filepaths
encodings = gwhap:::gwhapConfig$genmap_toy_reference_1000$encodings

ref1000_genetic_map = Genetic_Map(filepaths=filepaths, encodings=encodings)
ref1000_genetic_map = readData(ref1000_genetic_map)


snp_file=system.file("extdata","haplotypes.bgen.bgi", package="gwhap", mustWork=TRUE)
snp_physical_positions=sprintf("%s/ukb_",dirname(snp_file))
interpolated_map=create_augmented_genetic_map(snp_physical_positions=snp_physical_positions, genetic_map=ref1000_genetic_map, genetic_map_dir=genetic_map_dir, map_name='1000_genome', save_genetic_map=FALSE)

head(interpolated_map)
```



* Using UKB Biobank filenames format (bgen.bgi) and  Rutgers maps:

```{r}
filepaths = gwhap:::gwhapConfig$genmap_toy_rutger$filepaths
encodings = gwhap:::gwhapConfig$genmap_toy_rutger$encodings

rutger_genetic_map = Genetic_Map(filepaths=filepaths, encodings=encodings)
rutger_genetic_map = readData(rutger_genetic_map)

snp_physical_positions="inst/extdata/ukb_"
interpolated_map=create_augmented_genetic_map(snp_physical_positions=snp_physical_positions, genetic_map=rutger_genetic_map, genetic_map_dir=genetic_map_dir, map_name='rutgers', save_genetic_map=FALSE)

head(interpolated_map)
```


